forester/src/batch_processor/state.rs

```diff
@@ -15,77 +18,188 @@ use crate::{
     },
 };
 
+#[instrument(
+    level = "debug",
+    fields(
+        forester = %context.derivation,
+        epoch = %context.derivation,
+        merkle_tree = %context.merkle_tree,
+        output_queue = %context.output_queue,
+    ), skip(context, rpc))
+]
 pub(crate) async fn perform_append<R: RpcConnection, I: Indexer<R> + IndexerType<R>>(
     context: &BatchContext<R, I>,
     rpc: &mut R,
 ) -> Result<()> {
-    debug!("perform_append");
-    let instruction_data = create_append_batch_ix_data(
+    let instruction_data_vec = create_append_batch_ix_data(
         rpc,
         &mut *context.indexer.lock().await,
         context.merkle_tree,
         context.output_queue,
     )
     .await
-    .map_err(|e| BatchProcessError::InstructionData(e.to_string()))?;
-    let instruction = create_batch_append_instruction(
-        context.authority.pubkey(),
-        context.derivation,
-        context.merkle_tree,
-        context.output_queue,
-        context.epoch,
-        instruction_data
-            .try_to_vec()
-            .map_err(|e| BatchProcessError::InstructionData(e.to_string()))?,
+    .map_err(|e| {
+        error!("Failed to create append batch instruction data: {}", e);
+        BatchProcessError::InstructionData(e.to_string())
+    })?;
+
+    if instruction_data_vec.is_empty() {
+        debug!("No zkp batches to append");
+        return Ok(());
+    }
+
+    info!(
+        "Processing {} ZKP batch appends",
+        instruction_data_vec.len()
     );
 
-    rpc.create_and_send_transaction(
-        &[instruction],
-        &context.authority.pubkey(),
-        &[&context.authority],
-    )
-    .await?;
+    for (idx, instruction_data) in instruction_data_vec.iter().enumerate() {
```

Comment:
we could iterate over instruction_data_vec in chunks eg 8 (I am not sure how many we can fit into one tx).
```suggestion
    for (idx, instruction_data) in instruction_data_vec.iter().chunks(8).enumerate() {
```